[New] Authenticate with OAuth (#126)
Esri/arcgis-maps-sdk-swift-samples: Shared/Samples/Authenticate with OAuth/README.md

# Authenticate with OAuth

Authenticate with ArcGIS Online (or your own portal) using OAuth2 to access secured resources (such as private web maps or layers).

![Login screen](authenticate-with-oauth1.png)

> ![Map view after authentication](authenticate-with-oauth2.png)

Reviewer: Can you retake these screenshots so that they show the standard 9:41?

## Use case

Your app may need to access items that are only shared with authorized users. For example, your organization may host private data layers or feature services that are only accessible by verified users. You may also need to take advantage of premium ArcGIS Online services, such as geocoding or routing, that require a named user login.

## How to use the sample

> Upon launch, a web map containing premium content will load. You will be prompted by an ArcGIS Online login to view the private layers. Enter a user name and password for an ArcGIS Online named user account (such as your ArcGIS for Developers account). If you authenticate successfully, the traffic layer will display; otherwise, the map will contain only the public basemap layer.

Author: done

> Upon launch, a web map containing premium content will load. You will be prompted by an ArcGIS Online login to view the private layers. Enter a user name and password for an ArcGIS Online named user account (such as your ArcGIS for Developers account). If you authenticate successfully, the traffic layer will display; otherwise, the map will contain only the public basemap layer.

Reviewer 2: Maybe the change wasn't committed? I don't see it here. 🤔

> Upon launch, a web map containing premium content will load. You will be prompted by an ArcGIS Online login to view the private layers. Enter a user name and password for an ArcGIS Online named user account (such as your ArcGIS for Developers account). If you authenticate successfully, the traffic layer will display; otherwise, the map will contain only the public basemap layer.

Author: Changed readme in VS Code and did not hit the save 🤦‍♂️

> When you run the sample, the app will load a web map which contains premium content. You will be challenged for an ArcGIS Online login to view the private layers. Enter a user name and password for an ArcGIS Online named user account (such as your ArcGIS for Developers account). If you authenticate successfully, the traffic layer will display, otherwise the map will contain only the public basemap layer.

Reviewer: ```suggestion
Upon launch, a web map containing premium content will load. You will be prompted by an ArcGIS Online login to view the private layers. Enter a user name and password for an ArcGIS Online named user account (such as your ArcGIS for Developers account). If you authenticate successfully, the traffic layer will display; otherwise, the map will contain only the public basemap layer.
```
Just my opinions, feel free to pick and choose from my suggestions.

Author: I'm ok with the change but this is same as in iOS sdk sample.

## How it works

1. Create an `OAuthUserConfiguration` object specifying the portal URL, client ID, and redirect URL.
2. Create a toolkit component `Authenticator` object with OAuth user configuration.
3. Set authenticator object as ArcGIS and Network challenge handlers on authentication manager to handle authentication challenges.
4. Load a map with premium content requiring authentication.

## Relevant API

* AuthenticationManager
* Authenticator
* OAuthUserConfiguration
* Portal
* PortalItem

## Additional information

The workflow presented in this sample works for all SAML based enterprise (IWA, PKI, Okta, etc.) & social (Facebook, Google, etc.) identity providers for ArcGIS Online or Portal. For more information, see the topic [Set up enterprise logins](https://doc.arcgis.com/en/arcgis-online/administer/enterprise-logins.htm).

> For additional information on using OAuth in your app, see the topic [Authenticate with the API](https://developers.arcgis.com/documentation/core-concepts/security-and-authentication/mobile-and-native-user-logins/) in *Mobile and Native Named User Login*.

Reviewer: Is this the correct link? It seems to redirect me to https://developers.arcgis.com/documentation/mapping-apis-and-services/security/arcgis-identity/serverless-native-apps/

Author: The link is same as in iOS sdk sample.

Author: done
